crates/bevy_app/src/plugin.rs

```diff
@@ -56,19 +89,46 @@ use std::any::Any;
 /// # fn damp_flickering() {}
 /// ````
 pub trait Plugin: Downcast + Any + Send + Sync {
-    /// Configures the [`App`] to which this plugin is added.
-    fn build(&self, app: &mut App);
+    /// Returns required sub apps before finalizing this plugin.
+    fn require_sub_apps(&self) -> Vec<InternedAppLabel> {
+        Vec::new()
+    }
 
-    /// Has the plugin finished its setup? This can be useful for plugins that need something
-    /// asynchronous to happen before they can finish their setup, like the initialization of a renderer.
-    /// Once the plugin is ready, [`finish`](Plugin::finish) should be called.
-    fn ready(&self, _app: &App) -> bool {
+    /// Pre-configures the [`App`] to which this plugin is added. This is usually executed before
+    /// the event loop is started.
+    fn init(&self, _app: &mut App) {
+        // do nothing
+    }
+
+    /// Is the plugin ready to be built?
+    fn ready_to_build(&self, _app: &mut App) -> bool {
         true
     }
 
-    /// Finish adding this plugin to the [`App`], once all plugins registered are ready. This can
+    /// Builds the [`Plugin`] resources. This is usually executed inside an event loop.
+    fn build(&self, _app: &mut App) {
+        // do nothing
+    }
+
+    /// Is the plugin ready to be configured?
+    fn ready_to_configure(&self, _app: &mut App) -> bool {
+        true
+    }
+
+    /// Configures the [`App`] to which this plugin is added. This can
+    /// be useful for plugins that needs completing asynchronous configuration.
+    fn configure(&self, _app: &mut App) {
+        // do nothing
+    }
+
+    /// Is the plugin ready to be finalized?
+    fn ready_to_finalize(&self, _app: &mut App) -> bool {
+        true
+    }
+
+    /// Finalizes this plugin to the [`App`]. This can
     /// be useful for plugins that depends on another plugin asynchronous setup, like the renderer.
-    fn finish(&self, _app: &mut App) {
+    fn finalize(&self, _app: &mut App) {
```

Comment:
We should have a graceful migration from `finish` to finalize. Deprecate `finish` rather than immediately removing it, and default `finalize` to call `finish`